app/src/main/java/org/openobservatory/ooniprobe/activity/RunTestActivity.java

```diff
@@ -0,0 +1,236 @@
+            String mv = uri.getQueryParameter("mv");
+            if (mv != null){
+                if (versionCompare(BuildConfig.VERSION_NAME, mv) >= 0) {
+                    Set<String> parameters = uri.getQueryParameterNames();
+                    String tn = uri.getQueryParameter("tn");
+                    String ta = uri.getQueryParameter("ta");
+                    String td = uri.getQueryParameter("td");
+                    String test = NetworkMeasurement.getTestName(this, tn);
+                    if (test.length() > 0){
+                        test_name = tn;
+                        configureScreen(td, ta);
+                    }
+                    else {
+                        Alert.alertDialog(this, getString(R.string.invalid_parameter), getString(R.string.test_name) +  " : " + tn);
+                    }
+                }
+                else {
+                    Alert.alertDialogTwoButtons(this, getString(R.string.ooniprobe_outdate), getString(R.string.ooniprobe_outdate_msg), new DialogInterface.OnClickListener() {
+                        public void onClick(DialogInterface dialog, int id) {
+                            final String appPackageName = getPackageName(); // getPackageName() from Context or Activity object
+                            try {
+                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=" + appPackageName)));
+                            } catch (android.content.ActivityNotFoundException anfe) {
+                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=" + appPackageName)));
+                            }
+                        }
+                    });
+                }
+            }
+        }
+    }
+
+    public void configureScreen(String td, String ta){
+        TestData.getInstance(this, this).addObserver(this);
+         urls = (TextView) findViewById(R.id.urls);
+
+        title = (TextView) findViewById(R.id.run_test_message);
+        if (td != null)
+            title.setText(td);
+        else
+            title.setText(getString(R.string.run_test_message));
+
+        String test = NetworkMeasurement.getTestName(this, test_name);
+        testTitle = (TextView) findViewById(R.id.test_title);
+        testTitle.setText(test);
+
+        testImage = (ImageView) findViewById(R.id.test_logo);
+        testImage.setImageResource(NetworkMeasurement.getTestImageBig(test_name));
+
+        ArrayList<String> listItems = new ArrayList<>();
+        final ArrayList<String> urlItems = new ArrayList<>();
+        if (ta != null){
+            try {
+                JSONObject taJson = new JSONObject(ta);
+                JSONArray urls = taJson.getJSONArray("urls");
+                for (int i = 0; i < urls.length(); i++)
+                    urlItems.add(urls.getString(i));
+            } catch (JSONException e) {
+                e.printStackTrace();
```

Comment:
Do we want to continue processing in this case?

Reply:
What else do you want to do? You just don't add urls

Reply:
I don't know, perhaps do not run the tests because the JSON is malformed?

Reply:
the parameter name test name is already validated, do you think we should abort all because the urls are malformed?

Reply:
Can we add a check to validate the URLs (by parsing the JSON) when the test name parameter is validated? This way we could show an error as early on as possible.

Reply:
I think that if we are at this point of the code and the JSON is malformed, we should not proceed with running the test, but rather should show an error message of some sort and abort running the test.

Reply:
What should happen if the JSON is malformed? I can show a popup, and then?